File: tests/gpu/spy_district_test.py
"""THE SPY STANDS ON A DISTRICT — the GPU half.

    npm run seed && npm run export        # (once) writes seeder/worlds/
    python tests/gpu/spy_district_test.py

The TS twin is tests/cpu/espionage/spy-district.test.ts.

CIV6 (UnitOperations): every spy operation names a `TargetDistrict` — Steal
Tech Boost the Campus, Sabotage Production the Industrial Zone — or none,
the City Center. A spy occupies the district it works out of. No gate lane
runs a spy mission end to end, so these scenes are the evidence:
  1. the travel head offers the DISTRICT tiles of a revealed city, nearest
     first, and a jump lands the spy on the tile it named
  2. a mission is offered where its district stands under the spy — the
     Zone's on the Zone, the centre's on the centre — and the counterspy
     post on any district of an own city
  3. CIV6 (Surveillance): the post defends every district of its city, and
     works a level higher within one hex of where it stands; Polygraph reads
     the whole city
  4. CIV6 (Sabotage Production) pillages the Zone's BUILDINGS: the yield
     walk skips them, the queue offers the repair at a quarter of the price,
     the gold arm does not, and the repair completing clears the flag
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
sys.path.insert(0, str(Path(__file__).resolve().parent))
from core import BatchSim, load_rules, load_fixture, fixture_paths
from warmup import settle_all

B0 = 0
ROOT = Path(__file__).resolve().parent.parent.parent
RULES = json.loads((ROOT / "seeder" / "worlds" / "rules.json").read_text())
BLD = [b["id"] for b in RULES["buildings"]]


def fresh(rules, path, turns=30):
    sim = settle_all(BatchSim([load_fixture(path)], rules, device="cpu", dtype=torch.float64))
    for _ in range(turns):
        sim.step()
    return sim


def a_city(sim, row):
    live = sim.city_alive[B0, row].nonzero().flatten()
    assert live.numel(), f"row {row} holds no city"
    return int(live[0])


def give_district(sim, row, j, di, dist: int | None = None):
    """A COMPLETE district of type `di` on a free plot the city owns — at
    exactly `dist` from the centre when asked."""
    ctr = int(sim.city_center[B0, row, j])
    free = [t for t in range(sim.T)
            if int(sim.tile_seat[B0, t]) == row and int(sim.district[B0, t]) < 0
            and int(sim.built_wonder[B0, t]) < 0 and bool(sim.passable[B0, t]) and t != ctr
            and (dist is None or int(sim.pair_dist[ctr, t]) == dist)]
    assert free, f"the city owns no free plot at distance {dist}"
    t = free[0]
    sim.district[B0, t] = di
    sim.district_complete[B0, t] = True
    sim.district_pillaged[B0, t] = False
    sim.city_dist_tile[B0, row, j, di] = t
    sim._eff_version += 1
    sim._tile_owner_ver += 1
    return t


def spawn_spy(sim, row, tile):
    was = set(sim.major_unit_alive[B0].nonzero().flatten().tolist())
    sim._spawn_unit(row, torch.ones(1, dtype=torch.bool), torch.tensor([tile]),
                    torch.tensor([sim._spy_idx]))
    got = set(sim.major_unit_alive[B0].nonzero().flatten().tolist()) - was
    assert len(got) == 1, "the spawn found no slot"
    sim._gen_ver += 1
    return got.pop()


def rank_of(sim, row, slot):
    smap = sim._seat_slot_map(row)[B0]
    return int((smap == slot).nonzero(as_tuple=True)[0][0])


def order(sim, row, slot, col):
    smap = sim._seat_slot_map(row)[B0]
    acts = torch.full((1, smap.shape[0]), -1, dtype=torch.long)
    acts[0, rank_of(sim, row, slot)] = col
    sim.seat_ext[B0, row] = True
    sim._apply_seat_unit_actions(row, acts)


def missions(sim, row, slot):
    m = sim._seat_unit_mask(row)
    return m[B0, rank_of(sim, row, slot)][sim._A_SPY_MISSION:sim._A_SPY_MISSION + sim._n_spy_missions]


def pcol(sim, kind):
    rd = sim.rules_dev
    ec = list(rd.promo_classes).index("ESPIONAGE")
    k = sim._pk[kind]
    hit = [j for j in range(int(rd.promo_rows[ec]))
           if any(int(rd.promo_kind[ec, j, s]) == k for s in range(rd.promo_kind.shape[2]))]
    assert len(hit) == 1, f"{kind} names {len(hit)} columns"
    return hit[0]


def main() -> None:
    rules = load_rules()
    path = fixture_paths()[0]
    row, foe = 1, 2
    sim = fresh(rules, path)
    assert sim._spy_idx >= 0 and sim._iz_idx >= 0 and sim._campus_idx >= 0
    mine, theirs = a_city(sim, row), a_city(sim, foe)
    ctr_m = int(sim.city_center[B0, row, mine])
    ctr_t = int(sim.city_center[B0, foe, theirs])
    iz = give_district(sim, foe, theirs, sim._iz_idx, dist=1)
    far = give_district(sim, foe, theirs, sim._campus_idx, dist=2)
    sim.seat_explored[B0, row] = True

    # -- 1: the travel head names district tiles, nearest first -------------
    v = spawn_spy(sim, row, ctr_m)
    cols = sim._spy_destinations(row, torch.tensor([[v]]), torch.tensor([[ctr_m]]),
                                 torch.tensor([[sim._spy_idx]]))[B0, 0].tolist()
    live = [c for c in cols if c >= 0]
    assert ctr_t in live and iz in live and far in live, "a district tile of a revealed city is missing"
    assert ctr_m not in live, "the spy's own tile is offered"
    d = [int(sim.pair_dist[ctr_m, c]) for c in live]
    assert d == sorted(d), f"the head is not nearest-first: {d}"
    k = live.index(iz)
    order(sim, row, v, sim._A_SPY_TRAVEL + k)
    assert int(sim.unit_spy_target[B0, v]) == iz, "the jump did not name the Zone's tile"
    for _ in range(int(sim.unit_spy_turns[B0, v])):
        sim._tick_spies(row)
    assert int(sim.unit_tile[B0, v]) == iz, "the spy did not land on the Zone"
    hrow, hcol = sim._spy_here(torch.tensor([[iz]]))
    assert (int(hrow[0, 0]), int(hcol[0, 0])) == (foe, theirs), "the Zone is not read as its city's"
    print("  1 travel OK — district tiles offered nearest first, the spy lands on the one it named")

    # -- 2: the geometry of the mask ----------------------------------------
    mm = missions(sim, row, v)
    assert bool(mm[sim._spy_m_sabotage]), "Sabotage is not offered on the Zone"
    assert not bool(mm[sim._spy_m_sources]) and not bool(mm[sim._spy_m_unrest]), "a centre mission lit on the Zone"
    sim.unit_tile[B0, v] = ctr_t
    sim._gen_ver += 1
    mm = missions(sim, row, v)
    assert not bool(mm[sim._spy_m_sabotage]), "Sabotage is offered from the centre"
    assert bool(mm[sim._spy_m_unrest]), "a centre mission is not offered on the centre"
    assert not bool(mm[sim._spy_m_counterspy]), "the post is offered in a rival's city"
    own_c = give_district(sim, row, mine, sim._campus_idx)
    sim.unit_tile[B0, v] = own_c
    sim._gen_ver += 1
    mm = missions(sim, row, v)
    assert bool(mm[sim._spy_m_counterspy]), "the post is not offered on an own district"
    sim.unit_tile[B0, v] = ctr_m
    sim._gen_ver += 1
    assert bool(missions(sim, row, v)[sim._spy_m_counterspy]), "the post is not offered on the own centre"
    print("  2 mask OK — a mission where its district stands, the post on any own district")

    # -- 3: Surveillance's reach and Polygraph's whole city -----------------
    guard = spawn_spy(sim, foe, ctr_t)
    sim.unit_spy_mission[B0, guard] = sim._spy_m_counterspy
    base_iz = sim._counter_levels(B0, foe, theirs, iz)
    base_far = sim._counter_levels(B0, foe, theirs, far)
    sim.unit_promos[B0, guard] = 1 << pcol(sim, "SPY_SURVEIL")
    assert sim._counter_levels(B0, foe, theirs, iz) == base_iz + 1, "no +1 level within one hex of the post"
    assert sim._counter_levels(B0, foe, theirs, far) == base_far, "the +1 reached two hexes out"
    assert sim._counter_levels(B0, foe, theirs, ctr_t) == base_iz + 1, "the post's own tile is within reach"
    # the post GUARDS the district the intruder works from: its own tile
    # always, every district of the city with Surveillance
    sim.unit_promos[B0, guard] = 0
    assert sim._counterspies_guarding(B0, foe, theirs, iz).numel() == 0, "a post on the centre guarded the Zone unpromoted"
    assert sim._counterspies_guarding(B0, foe, theirs, ctr_t).tolist() == [guard]
    sim.unit_promos[B0, guard] = 1 << pcol(sim, "SPY_SURVEIL")
    assert sim._counterspies_guarding(B0, foe, theirs, iz).tolist() == [guard], "Surveillance did not extend the guard"
    assert sim._counterspies_guarding(B0, foe, theirs, far).tolist() == [guard]
    # Polygraph: the post reads anywhere in the city
    sim.unit_promos[B0, guard] = 1 << pcol(sim, "SPY_HOME_ENEMY_LEVEL")
    sim.unit_tile[B0, guard] = far
    sim._gen_ver += 1
    assert sim._counter_levels(B0, foe, theirs, iz) == base_iz + 1, "Polygraph on a district did not reach the city"
    sim.unit_alive[B0, guard] = False
    sim._gen_ver += 1
    print("  3 surveillance OK — every district guarded, +1 level within one hex; Polygraph city-wide")

    # -- 4: Sabotage pillages the Zone's buildings; the queue repairs them ---
    wk = int((sim._b_req_district == sim._iz_idx).nonzero(as_tuple=True)[0][0])
    sim.city_bldg[B0, foe, theirs, wk] = True
    # the repair is offered under the building's own unlock, like the build
    ut, uc = int(sim.rules_dev.b_unlock[wk]), int(sim.rules_dev.b_unlock_civic[wk])
    if ut >= 0:
        sim.civ_techs[B0, foe, ut] = True
    if uc >= 0:
        sim.civ_civics[B0, foe, uc] = True
    sim._bldg_version += 1
    sim._eff_version += 1
    yf = sim._seat_amenity(foe)[2][:, theirs:theirs + 1]
    prod_lit = float(sim._seat_city_walk(foe, theirs, amen_yf=yf)[B0, 0, 1])
    sim.unit_tile[B0, v] = iz
    sim.unit_spy_mission[B0, v] = sim._spy_idle
    sim._gen_ver += 1
    order(sim, row, v, sim._A_SPY_MISSION + sim._spy_m_sabotage)
    sim.rng_state[B0] = 7  # a draw that clears the success bar
    for _ in range(int(sim.unit_spy_turns[B0, v])):
        sim._tick_spies(row)
    assert bool(sim.city_bldg_pillaged[B0, foe, theirs, wk]), "Sabotage did not pillage the Workshop"
    assert not bool(sim.district_pillaged[B0, iz]), "Sabotage darkened the district"
    yf = sim._seat_amenity(foe)[2][:, theirs:theirs + 1]
    prod_dark = float(sim._seat_city_walk(foe, theirs, amen_yf=yf)[B0, 0, 1])
    assert prod_dark < prod_lit, f"the yield walk still pays the pillaged Workshop ({prod_lit} -> {prod_dark})"
    # the REPAIR: the building's own column, at a quarter of the price, from the queue alone
    assert bool(sim._seat_buildable(foe)[B0, theirs, wk]), "the pillaged Workshop is not offered for repair"
    assert not bool(sim._seat_buildable(foe, gold=True)[B0, theirs, wk]), "the gold arm sells a repair"
    full = float(sim.rules_dev.b_cost[wk])
    pct = float(sim.rules.pillage_building_repair_pct)
    want = round(full * pct / 100.0)
    got = float(sim._building_cost_in(foe, theirs, torch.tensor([wk]))[B0])
    assert got == want, f"the repair is priced {got}, want {want} ({pct}% of {full})"
    hit = torch.zeros(sim.B, dtype=torch.bool)
    hit[B0] = True
    sim._q_push(foe, theirs, hit, torch.full((sim.B,), wk, dtype=torch.long),
                torch.full((sim.B,), want, dtype=sim.city_cost.dtype))
    sim._seat_city_produce(foe, torch.full((sim.B,), theirs, dtype=torch.long), hit,
                           torch.full((sim.B,), want, dtype=torch.float64))
    assert not bool(sim.city_bldg_pillaged[B0, foe, theirs, wk]), "the repair completing did not clear the flag"
    assert bool(sim.city_bldg[B0, foe, theirs, wk]), "the repair lost the building"
    assert not bool(sim._seat_buildable(foe)[B0, theirs, wk]), "a repaired building is still offered"
    print(f"  4 sabotage OK — the Workshop pillaged and dark, repaired at {want:.0f} of {full:.0f}")

    print("BATTERY OK spy_district")


if __name__ == "__main__":
    main()

File: tests/cpu/espionage/spy-district.test.ts
/**
 * THE SPY STANDS ON A DISTRICT — the TS half (the GPU twin is
 * tests/gpu/spy_district_test.py).
 *
 * CIV6 (UnitOperations): every spy operation names a `TargetDistrict` — Steal
 * Tech Boost the Campus, Sabotage Production the Industrial Zone — or none,
 * the City Center. A spy occupies the district it works out of. No gate lane
 * runs a spy mission end to end, so these scenes are the evidence.
 */
import { describe, it, expect } from 'vitest';
import { makeMap, makeState, tileAtCoords, settleAt } from '../helpers';
import { spawnUnit } from '../../../cpu/core/units';
import { emptySeat, seatOf, setTileOwner } from '../../../cpu/core/seats';
import {
  spyDestinations, beginTravel, beginMission, missionOffered, tickSpies, spyCity, cityCounterLevels,
} from '../../../cpu/core/espionage';
import { spyHeldWith } from '../../../cpu/core/deals';
import { promoRows } from '../../../cpu/data/promotions';
import {
  SPY_UNIT, SPY_MISSIONS, SPY_M_SABOTAGE_PRODUCTION, SPY_M_STEAL_TECH_BOOST, SPY_M_FOMENT_UNREST,
  SPY_M_COUNTERSPY, SPY_M_GAIN_SOURCES, SPY_M_SIPHON_FUNDS, SPY_ESCAPE_ROUTES, SPY_SURVEILLANCE_REACH,
} from '../../../cpu/data/espionage';
import { buildingPillaged, cityBuildingYields, darkBuildings } from '../../../cpu/core/yields';
import { availableBuildings, goldPurchasableBuildings, buildingCostIn } from '../../../cpu/core/rules';
import { completeQueueItem } from '../../../cpu/core/production';
import { makeYieldCtx } from '../../../cpu/core/effects';
import { BUILDINGS } from '../../../cpu/data/buildings';
import { PILLAGE_BUILDING_REPAIR_PERCENT } from '../../../cpu/data/constants';
import { hexDistance } from '../../../world/hex';
import type { City, GameState, QueueItem } from '../../../cpu/core/types';

const WINS = 7;
const turnsOf = (m: number): number => SPY_MISSIONS[m]!.turns;
function spyBit(id: string): number {
  const k = promoRows('ESPIONAGE').findIndex((p) => p.id === id);
  expect(k).toBeGreaterThanOrEqual(0);
  return 1 << k;
}

/** Two majors, a city each, both spy-capable. */
function spyState() {
  const state = makeState(makeMap(24, 24));
  state.unitsMode = true;
  state.seats.push(emptySeat(1));
  const mine = settleAt(state, tileAtCoords(state.map, 4, 4).index, 0);
  const theirs = settleAt(state, tileAtCoords(state.map, 16, 16).index, 1);
  for (const s of state.seats) {
    s.treasury = 10_000;
    if (!s.research.civics.includes('DIPLOMATIC_SERVICE')) s.research.civics.push('DIPLOMATIC_SERVICE');
  }
  return { state, mine, theirs, me: seatOf(state, 0)!, them: seatOf(state, 1)! };
}

/** a COMPLETE district of `type` exactly `dist` tiles from the centre */
function districtAt(state: GameState, city: City, type: string, dist: number): number {
  const ctr = state.map.tiles[city.centerIndex];
  const t = state.map.tiles.find(
    (x) => !x.district && x.terrain !== 'OCEAN' && hexDistance(x.col, x.row, ctr.col, ctr.row) === dist,
  )!;
  setTileOwner(t, city.seat, city.id);
  t.district = type as never;
  t.districtComplete = true;
  city.districts.push({ type: type as never, tileIndex: t.index });
  return t.index;
}

function spyAt(state: GameState, seat: number, tile: number) {
  const u = spawnUnit(state, SPY_UNIT, tile, seat)!;
  u.tileIndex = tile;
  return u;
}

describe('the jump names a district tile', () => {
  it('offers every district of a revealed city, nearest first, and lands the spy on it', () => {
    const { state, mine, theirs, them } = spyState();
    const iz = districtAt(state, theirs, 'INDUSTRIAL_ZONE', 1);
    const campus = districtAt(state, theirs, 'CAMPUS', 2);
    const spy = spyAt(state, 0, mine.centerIndex);
    const dests = spyDestinations(state, spy);
    expect(dests).toContain(theirs.centerIndex);
    expect(dests).toContain(iz);
    expect(dests).toContain(campus);
    expect(dests).not.toContain(mine.centerIndex);
    const here = state.map.tiles[mine.centerIndex];
    const d = dests.map((t) => hexDistance(here.col, here.row, state.map.tiles[t].col, state.map.tiles[t].row));
    expect(d).toEqual([...d].sort((a, b) => a - b));

    expect(beginTravel(state, spy, campus)).toBe(true);
    expect(spy.spyTarget).toBe(campus);
    const turns = spy.spyTurns ?? 0;
    expect(turns).toBeGreaterThan(0);
    for (let i = 0; i < turns; i++) tickSpies(state, 0);
    expect(spy.tileIndex).toBe(campus);
    expect(spyCity(state, spy)?.city).toBe(theirs);
    // ...and the Campus offers the Campus' mission, not the centre's
    them.research.techs.push('MINING');
    expect(missionOffered(state, spy, SPY_M_STEAL_TECH_BOOST)).toBe(true);
    expect(missionOffered(state, spy, SPY_M_FOMENT_UNREST)).toBe(false);
    expect(missionOffered(state, spy, SPY_M_GAIN_SOURCES)).toBe(false);
    expect(missionOffered(state, spy, SPY_M_SABOTAGE_PRODUCTION)).toBe(false);
  });

  it('the counterspy post is offered on any district of an own city, and nowhere abroad', () => {
    const { state, mine, theirs } = spyState();
    const home = districtAt(state, mine, 'CAMPUS', 1);
    const away = districtAt(state, theirs, 'CAMPUS', 1);
    const spy = spyAt(state, 0, home);
    expect(missionOffered(state, spy, SPY_M_COUNTERSPY)).toBe(true);
    spy.tileIndex = mine.centerIndex;
    expect(missionOffered(state, spy, SPY_M_COUNTERSPY)).toBe(true);
    spy.tileIndex = away;
    expect(missionOffered(state, spy, SPY_M_COUNTERSPY)).toBe(false);
  });
});

describe('the counterspy defends the district it stands on', () => {
  it("SURVEILLANCE guards every district and works a level higher within one hex; POLYGRAPH reads the whole city", () => {
    const { state, theirs } = spyState();
    const near = districtAt(state, theirs, 'INDUSTRIAL_ZONE', 1);
    const far = districtAt(state, theirs, 'CAMPUS', 2);
    const guard = spyAt(state, 1, theirs.centerIndex);
    expect(beginMission(state, guard, SPY_M_COUNTERSPY)).toBe(true);
    expect(cityCounterLevels(state, theirs, near)).toBe(0);
    guard.promos = spyBit('SURVEILLANCE');
    expect(SPY_SURVEILLANCE_REACH).toBe(1);
    expect(cityCounterLevels(state, theirs, near)).toBe(1);
    expect(cityCounterLevels(state, theirs, theirs.centerIndex)).toBe(1);
    expect(cityCounterLevels(state, theirs, far)).toBe(0);
    // Polygraph counts from any district of the city
    guard.promos = spyBit('POLYGRAPH');
    guard.tileIndex = far;
    expect(cityCounterLevels(state, theirs, near)).toBe(1);
  });

  it('a post on the centre catches nobody on the Hub — until Surveillance extends it', () => {
    // the catch is pinned certain and every escape shut, so whether the post
    // GUARDS the Hub decides the whole outcome
    const row = SPY_MISSIONS[SPY_M_SIPHON_FUNDS] as { successPct?: number };
    const saved = row.successPct;
    const rates = SPY_ESCAPE_ROUTES.map((r) => r.basePct);
    row.successPct = -1000;
    for (const r of SPY_ESCAPE_ROUTES) (r as { basePct: number }).basePct = -1000;
    try {
      const run = (surveil: boolean): boolean => {
        for (let seed = 1; seed < 200; seed++) {
          const { state, theirs } = spyState();
          const hub = districtAt(state, theirs, 'COMMERCIAL_HUB', 1);
          const guard = spyAt(state, 1, theirs.centerIndex);
          if (surveil) guard.promos = spyBit('SURVEILLANCE');
          expect(beginMission(state, guard, SPY_M_COUNTERSPY)).toBe(true);
          const spy = spyAt(state, 0, hub);
          state.rngState = seed;
          expect(beginMission(state, spy, SPY_M_SIPHON_FUNDS)).toBe(true);
          for (let i = 0; i < turnsOf(SPY_M_SIPHON_FUNDS); i++) tickSpies(state, 0);
          if (spyHeldWith(state, 0, 1) === 1 && guard.spyLevel === 1) return true;
        }
        return false;
      };
      expect(run(false)).toBe(false);
      expect(run(true)).toBe(true);
    } finally {
      row.successPct = saved;
      SPY_ESCAPE_ROUTES.forEach((r, i) => { (r as { basePct: number }).basePct = rates[i]; });
    }
  });
});

describe('Sabotage Production pillages the buildings', () => {
  it('marks the Zone\'s buildings, the yield walk skips them, and the queue repairs them at a quarter', () => {
    const { state, theirs, them } = spyState();
    const iz = districtAt(state, theirs, 'INDUSTRIAL_ZONE', 1);
    theirs.buildings.push('WORKSHOP');
    them.research.techs.push('APPRENTICESHIP');
    const ctx = makeYieldCtx(state, theirs.seat);
    const lit = cityBuildingYields(ctx, theirs).production;
    const spy = spyAt(state, 0, iz);
    state.rngState = WINS;
    expect(beginMission(state, spy, SPY_M_SABOTAGE_PRODUCTION)).toBe(true);
    for (let i = 0; i < turnsOf(SPY_M_SABOTAGE_PRODUCTION); i++) tickSpies(state, 0);
    expect(buildingPillaged(theirs, 'WORKSHOP')).toBe(true);
    expect(state.map.tiles[iz].districtPillaged).toBe(false);
    expect(darkBuildings(state.map, theirs).has('WORKSHOP')).toBe(true);
    expect(cityBuildingYields(ctx, theirs).production).toBe(lit - (BUILDINGS.WORKSHOP.yields?.production ?? 0));

    // THE REPAIR: the building's own column, at PILLAGE_BUILDING_REPAIR_PERCENT
    // of its price, from the queue alone
    expect(availableBuildings(state, theirs).some((b) => b.id === 'WORKSHOP')).toBe(true);
    expect(goldPurchasableBuildings(state, theirs).some((b) => b.id === 'WORKSHOP')).toBe(false);
    const price = buildingCostIn(state, theirs, 'WORKSHOP');
    expect(price).toBe(Math.round((BUILDINGS.WORKSHOP.cost * PILLAGE_BUILDING_REPAIR_PERCENT) / 100));
    const item: QueueItem = { kind: 'building', building: 'WORKSHOP', progress: price };
    completeQueueItem(state, theirs, item, price);
    expect(buildingPillaged(theirs, 'WORKSHOP')).toBe(false);
    expect(theirs.buildings.filter((b) => b === 'WORKSHOP')).toHaveLength(1);
    expect(availableBuildings(state, theirs).some((b) => b.id === 'WORKSHOP')).toBe(false);
    expect(buildingCostIn(state, theirs, 'WORKSHOP')).toBe(BUILDINGS.WORKSHOP.cost);
    expect(cityBuildingYields(ctx, theirs).production).toBe(lit);
  });
});
